crates/iota-indexer/migrations/pg/2025-03-13-081550_insertion_order/up.sql

```diff
@@ -0,0 +1,16 @@
+-- Insertion order number that each transaction (either optimistic or
+-- checkpointed) is assigned when being indexed. It provides common
+-- ordering for optimistic and checkpointed transactions, whereas
+-- `tx_sequence_number` provides ordering only for checkpointed transactions.
+--
+-- In case of missing digests, the `tx_digests` table is used a fallback
+-- to resolve the transaction order. This is ok because optimistic transactions
+-- will be inserted only after creation of this table.
+CREATE SEQUENCE tx_insertion_order_seq;
+CREATE TABLE tx_insertion_order (
+    tx_digest                   BYTEA        PRIMARY KEY,
+    insertion_order             BIGINT       NOT NULL DEFAULT nextval('tx_insertion_order_seq')
+);
+ALTER SEQUENCE tx_insertion_order_seq OWNED BY tx_insertion_order.insertion_order;
+SELECT setval('tx_insertion_order_seq', (SELECT MAX(tx_sequence_number) FROM tx_digests));
```

Comment:
Could it happen in a very special case `tx_digests` doesn't contain any rows and it would return `NULL` in that case? Maybe something like this would help?

`SELECT setval('tx_insertion_order_seq', COALESCE((SELECT MAX(tx_sequence_number) FROM tx_digests), 0));`

Reply:
Good point yes. This may very well happen when we reset the database. Tested already with scenario 3 (see description). In this case the the `SELECT setval` itself returns NULL, so this is a no-op.